### PR TITLE
Rx rewrite

### PR DESCRIPTION
**Pull request by JakeGinnivan**
Builds on #72 but takes it a step or 5 (eventually) further

Outstanding issues:

 - [x] Message timeout needs to reset delay when message is modified

### CODE REVIEW

File: src/Carnac.Tests/MessageProviderFacts.cs [hunk 1]
```diff
@@ -61,12 +67,12 @@ public void recognises_shortcuts()
         }
 
         [Fact]
-        public void does_not_show_shortcut_name_on_partial_match()
+        public void does_not_show_key_press_on_partial_match()
```
Comment (JakeGinnivan) [inline src/Carnac.Tests/MessageProviderFacts.cs]: This is the test I changed the behaviour of.

Before if I was in visual studio and pressed `ctrl +r` that would appear in carnac right away. But because it is part of a chord shortcut I thought it might make sense to not show it until you completed/broke the chord.

What do you think @shiftkey

File: src/Carnac.Logic/MessageProvider.cs [hunk 2]
```diff
@@ -1,143 +1,49 @@
...
+        {
+            /*
+            shortcut Acc stream:
+               - ! before item means HasCompletedValue is false
+               - [1 & 2] means multiple messages are returned from get messages (1 and 2 in this case), others are a single message returned
+
+            message merger:
+               - * before items indicates the previous message has been modified (key has been merged into acc), otherwise new acc is created
+
+            keystream   :  a---b---ctrl+r----ctrl+r----------ctrl+r----a--------------↓---↓
```
Comment (JakeGinnivan) [inline src/Carnac.Logic/MessageProvider.cs]: Marble diagram ftw :)

**Comment by JakeGinnivan** [comment]
@shiftkey ready to go

File: src/Carnac.Logic/Models/Processinfo.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+namespace Carnac.Logic.Models
+{
+    public class Processinfo
```
Comment (shiftkey) [inline src/Carnac.Logic/Models/Processinfo.cs]: s/info/Info/g

File: src/Carnac.Logic/MessageMerger.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            if (ShouldCreateNewMessage(acc, key))
+                return key;
+
+            return acc.Merge(key);
+        }
+
+        bool ShouldCreateNewMessage(Message acc, Message key)
+        {
+            oneSecond = TimeSpan.FromSeconds(1);
```
Comment (shiftkey) [inline src/Carnac.Logic/MessageMerger.cs]: you can initialize this when you initialize the field

File: src/Carnac.Logic/MessageMerger.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
+using System.ComponentModel.Composition;
+using Carnac.Logic.Models;
+
+namespace Carnac.Logic
+{
+    [Export(typeof(IMessageMerger))]
+    public class MessageMerger : IMessageMerger
+    {
+        TimeSpan oneSecond;
```
Comment (shiftkey) [inline src/Carnac.Logic/MessageMerger.cs]: `static readonly TimeSpan OneSecond = TimeSpan.FromSeconds(1);`

File: src/Carnac.Logic/MessageMerger.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+        public Message MergeIfNeeded(Message acc, Message key)
+        {
+            if (ShouldCreateNewMessage(acc, key))
+                return key;
+
+            return acc.Merge(key);
+        }
+
+        bool ShouldCreateNewMessage(Message acc, Message key)
```
Comment (shiftkey) [inline src/Carnac.Logic/MessageMerger.cs]: This can then become a static method...

File: src/Carnac.Logic/ShortcutProvider.cs [hunk 7]
```diff
@@ -87,19 +87,13 @@ private KeyPressDefinition GetKeyPressDefintion(string combo)
...
-            {
-                var match = shortcut.GetShortcutsMatching(keyPresses);
-                if (match.Count() > 0)
-                    return match;
-            }
-
-            return Enumerable.Empty<KeyShortcut>();
+            var processName = keys.Process.ProcessName;
+            return shortcuts
+                .Where(s => (s.Process == processName) || string.IsNullOrWhiteSpace(s.Process))
```
Comment (shiftkey) [inline src/Carnac.Logic/ShortcutProvider.cs]: Drop the braces around `(s.Process == processName)` - they don't add anything

File: src/Carnac.Logic/ShortcutCollection.cs [hunk 8]
```diff
@@ -8,12 +8,12 @@ namespace Carnac.Logic
...
         public string Group { get; set; }
+
         public string Process { get; set; }
 
-        public IEnumerable<KeyShortcut> GetShortcutsMatching(IEnumerable<KeyPress> keys)
+        public KeyShortcut[] GetShortcutsMatching(IEnumerable<KeyPress> keys)
         {
-            var matches = this.Where(s => s.StartsWith(keys));
-            return matches;
+            return this.Where(s => s.StartsWith(keys)).ToArray();
```
Comment (shiftkey) [inline src/Carnac.Logic/ShortcutCollection.cs]: Is there a particular reason here why you favour `.ToArray()` over `.ToList()` here and in other places?

File: src/Carnac.Logic/ShortcutProvider.cs (hunk 7 above)
Comment (shiftkey) [inline src/Carnac.Logic/ShortcutProvider.cs]: Actually, should you care about case-sensitive comparisons here?

File: src/Carnac.Tests/MessageProviderFacts.cs (hunk 1 above)
Comment (shiftkey) [inline src/Carnac.Tests/MessageProviderFacts.cs]: I'll allow this, but I'd also love to wrap my head around the new chord changes and how it feels in the various apps (not just VS) so I might come back to this with better feedback.

File: src/Carnac.sln.DotSettings [hunk 9]
```diff
@@ -1,4 +1,7 @@
 ﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
+	<s:String x:Key="/Default/CodeInspection/Highlighting/InspectionSeverities/=AutoPropertyCanBeMadeGetOnly_002ELocal/@EntryIndexedValue">DO_NOT_SHOW</s:String>
```
Comment (shiftkey) [inline src/Carnac.sln.DotSettings]: I like R# being noisy. Convince me why we hide these.

File: src/Carnac/ConcurrencyService.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        readonly IScheduler uiScheduler;
+        readonly IScheduler defaultScheduler;
+
+        public ConcurrencyService()
+        {
+            uiScheduler = new SynchronizationContextScheduler(SynchronizationContext.Current);
+            defaultScheduler = Scheduler.Default;
+        }
+
+        public IScheduler UiScheduler
```
Comment (shiftkey) [inline src/Carnac/ConcurrencyService.cs]: I like `DispatcherScheduler` (Rx) or `MainThreadScheduler` (ReactiveUI) over this name

File: src/Carnac/KeysController.cs [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+using Carnac.Logic.Models;
+
+namespace Carnac
+{
+    public class KeysController : IDisposable
+    {
+        readonly ObservableCollection<Message> keys;
+        readonly IMessageProvider messageProvider;
+        readonly IKeyProvider keyProvider;
+        readonly TimeSpan fiveseconds = TimeSpan.FromSeconds(5);
```
Comment (shiftkey) [inline src/Carnac/KeysController.cs]: `fiveSeconds`

File: src/Carnac/KeysController.cs [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        readonly TimeSpan onesecond = TimeSpan.FromSeconds(1);
```
Comment (shiftkey) [inline src/Carnac/KeysController.cs]: `oneSecond` because we're not savages

File: src/Carnac/KeysController.cs [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+                    case ActionType.Remove:
+                        keys.Remove(action.Item1);
+                        break;
+                    case ActionType.FadeOut:
+                        action.Item1.IsDeleting = true;
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException();
+                }
+            }, ex =>
```
Comment (shiftkey) [inline src/Carnac/KeysController.cs]: Swallowing the errors feels bad, do we have any sort of logging at the moment to capture this?

File: src/Carnac/Views/ShellView.xaml [hunk 14]
```diff
@@ -6,11 +6,12 @@
...
                       xmlns:views="clr-namespace:Carnac.Views"
                       xmlns:xctk="http://schemas.xceed.com/wpf/xaml/toolkit" x:Class="Carnac.Views.ShellView"
-                      Width="610" Height="420" Icon="../icon.ico" ShowTitleBar="False"
+                      Width="610" Height="420" Icon="../icon.ico" 
                       Foreground="{DynamicResource BlackBrush}"
                       d:DataContext="{d:DesignInstance viewModels:ShellViewModel}" mc:Ignorable="d"
-                      ShowMinButton="False" ShowMaxRestoreButton="False"
-                      SaveWindowPosition="True">
+                      >
+    <!--ShowTitleBar="False" ShowMinButton="False" ShowMaxRestoreButton="False"
```
Comment (shiftkey) [inline src/Carnac/Views/ShellView.xaml]: :fire: this commented code

File: src/KeyStreamCapture/KeyStreamCapture.csproj [hunk 15]
```diff
@@ -11,7 +11,8 @@
     <RootNamespace>KeyStreamCapture</RootNamespace>
     <AssemblyName>KeyStreamCapture</AssemblyName>
     <TargetFrameworkVersion>v4.0</TargetFrameworkVersion>
-    <TargetFrameworkProfile>Client</TargetFrameworkProfile>
```
Comment (shiftkey) [inline src/KeyStreamCapture/KeyStreamCapture.csproj]: Can we back out from this change? It seems unnecessary to the overall PR (and might impact the installer)

**Comment by shiftkey** [comment]
Ok, I think I'm done. A few questions and little things...

File: src/Carnac.sln.DotSettings (hunk 9 above)
Comment (JakeGinnivan) [inline src/Carnac.sln.DotSettings]: These are roslyn fixes

Comment (shiftkey) [inline src/Carnac.sln.DotSettings]: Sure, just various feels about the additional "noise" in this PR. Would love it to go, but won't block it.

Comment (JakeGinnivan) [inline src/Carnac.sln.DotSettings]: Yeah. Just at different points r# generated code which would only compile on roslyn, could always pull it out into a separate PR but didn't see the point

File: src/Carnac.Tests/MessageProviderFacts.cs (hunk 1 above)
Comment (JakeGinnivan) [inline src/Carnac.Tests/MessageProviderFacts.cs]: Not too many apps do chord style shortcuts. Also feels weird to show 'ctrl+r' which does nothing because the app is waiting for the second half. Hiding incomplete shortcuts feels right

File: src/Carnac.Logic/ShortcutProvider.cs (hunk 7 above)
Comment (JakeGinnivan) [inline src/Carnac.Logic/ShortcutProvider.cs]: Case sensitive should be fine as it is process name which wouldn't change

File: src/Carnac.Logic/ShortcutCollection.cs (hunk 8 above)
Comment (JakeGinnivan) [inline src/Carnac.Logic/ShortcutCollection.cs]: Ideally I would to ToReadOnlyList() rather than array. But no immutable collections so array feels closer than list to that.

File: src/Carnac/KeysController.cs (hunk 13 above)
Comment (JakeGinnivan) [inline src/Carnac/KeysController.cs]: Never had any to start with. Just putting this here to show it. Maybe create another task and come up with a strategy

File: src/Carnac.Logic/ShortcutCollection.cs (hunk 8 above)
Comment (shiftkey) [inline src/Carnac.Logic/ShortcutCollection.cs]: Arrays are still mutable, but using a `List` here is slightly smarter in terms of memory allocation (because it doesn't attempt to trim the backing store to the length of the array).

File: src/Carnac/Views/ShellView.xaml (hunk 14 above)
Comment (JakeGinnivan) [inline src/Carnac/Views/ShellView.xaml]: Fixed in the next PR. For some reason the preferences window would not show up for me, I changed it to a normal window. These are MahApps.Metro properies, when I fixed it to use the MetroWindow again in the UI related PR I put these back.

File: src/KeyStreamCapture/KeyStreamCapture.csproj (hunk 15 above)
Comment (JakeGinnivan) [inline src/KeyStreamCapture/KeyStreamCapture.csproj]: Everything else is .net 4.0 and I had compilation issues because this was client profile referencing full framework stuff.

Comment (JakeGinnivan) [inline src/KeyStreamCapture/KeyStreamCapture.csproj]: Also this is only used to generate code for tests

File: src/Carnac/KeysController.cs (hunk 13 above)
Comment (shiftkey) [inline src/Carnac/KeysController.cs]: :thumbsup:

File: src/KeyStreamCapture/KeyStreamCapture.csproj (hunk 15 above)
Comment (shiftkey) [inline src/KeyStreamCapture/KeyStreamCapture.csproj]: :ok_hand:

**Comment by JakeGinnivan** [comment]
All comments addressed

File: src/Carnac/Views/ShellView.xaml (hunk 14 above)
Comment (shiftkey) [inline src/Carnac/Views/ShellView.xaml]: :ok_hand:

**Comment by shiftkey** [comment]
:shipit:
